### PR TITLE
APM-663: revision tree memory tracking

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Implements https://arangodb.atlassian.net/browse/APM-663

Tracks memory used for buffered updates to all revision trees in a new metric `arangodb_revision_tree_buffered_memory_usage`.

This PR also tries to optimize a problem with constant hibernation and resurrection patterns for the same tree. This was been found in reality by observing the hibernation and resurrection patterns of collections that is seldomly written into. For example, the `_statistics` system collection is written into only every ~10 seconds. Previously hibernation could happen randomly after a write to the collection, and then the next write would resurrect it. 
To break out of recurring unnecessary hibernation/resurrection cycles for the same tree, we now track the time of the last hibernation attempt for each revision tree, and do not attempt hibernation if the last attempt was less than a minute ago.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: -
  - [ ] Backport for 3.10: -
  - [ ] Backport for 3.9: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/APM-663
- [ ] Design document:

### CODE REVIEW

**Review by nerpaula — APPROVED** [review]
LGTM

**Review by markuspf — CHANGES_REQUESTED** [review]
Maybe the description should also mention that you introduce the `lastCompressAttempt` (and what for).

**Comment by jsteemann** [comment]
> Maybe the description should also mention that you introduce the `lastCompressAttempt` (and what for).

True. I added that later and forgot to mention it. Will add it.

File: Documentation/Metrics/arangodb_revision_tree_buffered_memory_usage.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+name: arangodb_revision_tree_buffered_memory_usage
+introducedIn: "3.12.0"
+help: |
+  Total memory usage of buffered updates for all revision trees.
+unit: bytes
+type: gauge
+category: Replication
+complexity: simple
```
Comment (maierlars) [inline Documentation/Metrics/arangodb_revision_tree_buffered_memory_usage.yaml:8]: Is this really simple, i.e. it should be visible on the most simple dashboard? I would go for `advanced`.

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 2]
```diff
@@ -1261,17 +1267,24 @@ void RocksDBEngine::trackRevisionTreeResurrection() noexcept {
 
 void RocksDBEngine::trackRevisionTreeMemoryIncrease(
     std::uint64_t value) noexcept {
-  if (value != 0) {
```
Comment (markuspf) [inline arangod/RocksDBEngine/RocksDBEngine.cpp:1264]: Do you by chance remember why this was here?

File: arangod/RocksDBEngine/RocksDBMetaCollection.h [hunk 3]
```diff
@@ -165,6 +170,15 @@ class RocksDBMetaCollection : public PhysicalCollection {
           std::unique_ptr<containers::RevisionTree>,
           std::unique_lock<std::mutex>& lock)> const& callback);
 
+  // used for calculating memory usage in _revisionsBufferedMemoryUsage
+  // approximate memory usage for top-level items
+  static constexpr uint64_t bufferedEntrySize() { return 40; }
```
Comment (markuspf) [inline arangod/RocksDBEngine/RocksDBMetaCollection.h:175]: How did this `40` get chosen?

File: tests/js/client/shell/shell-revision-tree-tracking-noncluster.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      const n = 10000;
+
+      let c = db._create(cn);
+      for (let i = 0; i < n; ++i) {
+        c.insert({value: i});
+      }
+
+      // must have more memory allocated for the documents
+      let metric = getMetric();
+      assertTrue(metric >= initial + n * (8 + 32), { metric, initial });
```
Comment (markuspf) [inline tests/js/client/shell/shell-revision-tree-tracking-noncluster.js:106]: Is this the same 40 as above?

**Review by markuspf — APPROVED** [review]
From what I can tell this looks good to me;

I am queasy about how to ascertain that all places that allocate/deallocate memory are accounted for: since I don't know the code in depth, the only way for me to tell is to learn this whole subsystem. That'd probably take too much time. Maybe we could have a quick zoom so I can get a high level view of what's happening?

File: arangod/RocksDBEngine/RocksDBEngine.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBEngine.cpp:1264]: It was there only to save an atomic write in case it wouldn't be necessary.

**Pull request merged**